packages/flutter/lib/src/widgets/widget_inspector.dart

```diff
@@ -0,0 +1,444 @@
+  }
+
+  SemanticsHandle _semanticsHandle;
+
+  @override
+  void dispose() {
+    _semanticsHandle.dispose();
+    _semanticsHandle = null;
+    super.dispose();
+  }
+
+  int generation = 0;
+
+  void _didUpdateSemantics() {
+    generation += 1;
+    notifyListeners();
+  }
+}
+
+const TextStyle _messageStyle = const TextStyle(
+    color: const Color(0xFFFFFFFF), fontSize: 10.0, height: 1.2);
+final int _kMaxTooltipLines = 5;
+const Color _kTooltipBackgroundColor = const Color.fromARGB(230, 60, 60, 60);
+
+void _paintDescription(Canvas canvas, RenderBox render, Offset target,
+    double verticalOffset, Size size, Rect targetRect) {
+  // TODO(jacobr): craft a better description message.
+  final String message = '$render\n\n'
+      '${render.debugCreator}';
+  canvas.save();
+  final TextPainter textPainter = new TextPainter()
+    ..maxLines = _kMaxTooltipLines
+    ..ellipsis = '...'
+    ..text = new TextSpan(style: _messageStyle, text: message)
+    ..layout(
+        maxWidth: size.width - 2 * (_kScreenEdgeMargin + _kTooltipPadding));
+
+  final Size tooltipSize =
+      textPainter.size + const Offset(_kTooltipPadding * 2, _kTooltipPadding * 2);
+  final _TooltipPositionCalculator calc = new _TooltipPositionCalculator(
+      target: target,
+      verticalOffset: verticalOffset,
+      preferBelow: false,
+      size: size,
+      childSize: tooltipSize);
+  final Offset tipOffset = calc.childPosition;
+
+  final Paint tooltipBackground = new Paint()
+    ..style = PaintingStyle.fill
+    ..color = _kTooltipBackgroundColor;
+  canvas.drawRect(
+      new Rect.fromPoints(tipOffset,
+          tipOffset.translate(tooltipSize.width, tooltipSize.height)),
+      tooltipBackground);
+
+  double wedgeY = tipOffset.dy;
+  if (!calc.tooltipBelow) {
+    wedgeY += tooltipSize.height;
+  }
+  final double wedgeSize = _kTooltipPadding * 2;
+  double wedgeX = math.max(tipOffset.dx, target.dx) + wedgeSize * 2;
+  wedgeX = math.min(wedgeX, tipOffset.dx + tooltipSize.width - wedgeSize * 2);
+  canvas.drawPath(
+      new Path()
+        ..addPolygon(<Offset>[
+          new Offset(wedgeX - wedgeSize, wedgeY),
+          new Offset(wedgeX + wedgeSize, wedgeY),
+          new Offset(
+              wedgeX, wedgeY + (calc.tooltipBelow ? -wedgeSize : wedgeSize)),
+        ], true),
+      tooltipBackground);
+
+  textPainter.paint(
+      canvas, tipOffset + const Offset(_kTooltipPadding, _kTooltipPadding));
+  canvas.restore();
+}
+
+final Color _kHighlightedRenderObjectFillColor =
+    const Color.fromARGB(128, 128, 128, 255);
+final Color _kHighlightedRenderObjectBorderColor =
+    const Color.fromARGB(128, 64, 64, 128);
+
+class _WidgetInspectorPainter extends CustomPainter {
+  const _WidgetInspectorPainter(this.owner, this.generation, this.selected);
+
+  final PipelineOwner owner;
+  final int generation;
+  final RenderBox selected;
+
+  bool get isSelectionActive => selected != null && selected.attached;
```

Comment:
if selected is not attached we should make a reasonable effort to see if there is an attached RenderBox that matches... e.g. same key or perhaps same class and path don't the render tree. Would be nice if you could scroll an element off screen and then back on and have the selection redisplay. Similarly, would be nice to continue showing the selection after as wide a range of hot reload edits as possible.